[qt] [tests] Clarify address book error messages, add tests

jamesob opened the pull request:
Addresses https://github.com/bitcoin/bitcoin/issues/12796.

When a user attempts to add to the address book a sending address which is already present as a receiving address, they're presented with a confusing error indicating the address is already present in the book, despite the fact that this row is currently invisible. 

![selection_011](https://user-images.githubusercontent.com/73197/38096704-8a2948d2-3341-11e8-9632-7d563201f28c.jpg)

This change adds a more specific error message indicating its existence as a receiving address (as discussed in the linked issue).

![selection_016](https://user-images.githubusercontent.com/73197/38198467-fa26164e-365a-11e8-8fc5-ddab9caf2fbd.jpg)

This change also adds some tests exercising use of the address book via QT. Adding so much test code for such a trivial change may seem weird, but it's my hope that this will make further test-writing for address book usage (and other QT features) more approachable.

jamesob committed:
[tests] [qt] Introduce qt/test/util with a generalized ConfirmMessage

ConfirmMessage is reused in future tests apart from its single usage here.

jamesob committed:
[tests] [qt] Add tests for address book manipulation via EditAddressDialog

Also modifies corresponding QT code to allow for use within test cases.

jamesob commented on src/qt/editaddressdialog.cpp:
```diff
@@ -129,6 +129,18 @@ void EditAddressDialog::accept()
+
+    if (bad_addr_purpose == "receive" &&
+            (mode == NewSendingAddress || mode == EditSendingAddress)) {
+        fmt_str = "Receiving address \"%1\" cannot be added as a sending address.";
```
Yeah, that's a good idea. Will add.

jamesob commented:
Incorporated @ryanofsky's feedback and rebased ([diff](https://gist.github.com/jamesob/6bd9721e11431d636fd2b9f6399a45ef)).

promag commented:
Concept ACK. Will review later.

jamesob commented:
Fixed Travis failures (unchecked `db_cxx.h` include, whitespace) and rebased.

jnewbery commented on src/qt/test/addressbooktests.cpp:
```diff
@@ -0,0 +1,140 @@
+ *
+ * There are three cases tested:
+ *
+ *   - new_address: a new address would should add as a send address successfully.
```
`would should`

jamesob replied:
Oops, fixed.

jnewbery reviewed (commented):
Tested ACK a5406962daa18a20e21eda8b578bccfad1630fd4

Looks great. Thanks for the new tests!

jamesob commented:
Thanks for the review, @jnewbery. Typo fixed and rebased.
```diff
 $ diff -u <(git diff master...a540696) <(git diff master...2d15e9c)

--- /proc/self/fd/11    2018-04-02 11:30:04.855218939 -0400
+++ /proc/self/fd/12    2018-04-02 11:30:04.855218939 -0400
@@ -192,7 +192,7 @@
      Mode mode;
 diff --git a/src/qt/test/addressbooktests.cpp b/src/qt/test/addressbooktests.cpp
 new file mode 100644
-index 0000000000..bc9ace14d7
+index 0000000000..84c765e48a
 --- /dev/null
 +++ b/src/qt/test/addressbooktests.cpp
 @@ -0,0 +1,140 @@
@@ -243,7 +243,7 @@
 + *
 + * There are three cases tested:
 + *
-+ *   - new_address: a new address would should add as a send address successfully.
++ *   - new_address: a new address which should add as a send address successfully.
 + *   - existing_s_address: an existing sending address which won't add successfully.
 + *   - existing_r_address: an existing receiving address which won't add successfully.
 + *
```

jnewbery commented:
reACK 2d15e9ce7d2ca4cc48ce0010e3c1c227a81f3b78

jamesob commented:
Rebased, conflicts from the #10244 merge resolved.

jnewbery commented on src/interfaces/wallet.h:
```diff
@@ -95,7 +95,8 @@ class Wallet
         std::string* name = nullptr,
-        isminetype* is_mine = nullptr) = 0;
+        isminetype* is_mine = nullptr,
+        std::string* purpose = nullptr) = 0;
```
This list of optional arguments slightly scares me. How much diff would it be to remove the defaults and change all the places this is called to pass in nullptr?

jamesob replied:
There are ~9 replacements this'd entail, so not too bad. Willfix.

jnewbery commented on src/qt/test/addressbooktests.cpp:
```diff
@@ -0,0 +1,144 @@
+    vpwallets.insert(vpwallets.begin(), &wallet);
+    WalletModel walletModel(std::move(node->getWallets()[0]), *node, platformStyle.get(), &optionsModel);
+    vpwallets.erase(vpwallets.begin());
+    AddressTableModel* addressTableModel = walletModel.getAddressTableModel();
```
unused?

```
qt/test/addressbooktests.cpp: In function ‘void {anonymous}::TestAddAddressesToSendBook()’:
qt/test/addressbooktests.cpp:109:24: warning: unused variable ‘addressTableModel’ [-Wunused-variable]
     AddressTableModel* addressTableModel = walletModel.getAddressTableModel();
                        ^
```

jamesob replied:
Good catch, thanks.

jnewbery commented on src/Makefile.qttest.include:
```diff
@@ -38,11 +41,13 @@ qt_test_test_bitcoin_qt_SOURCES = \
   qt/test/rpcnestedtests.cpp \
   qt/test/test_main.cpp \
   qt/test/uritests.cpp \
+  qt/test/util.cpp \
```
This line is added in the wrong commit (commit _Add tests for address book manipulation via EditAddressDialog_ instead of commit _Introduce qt/test/util with a generalized ConfirmMessage_). The intermediate commit fails linking.

jamesob replied:
Thanks for spotting this.

jamesob committed:
Add purpose arg to Wallet::getAddress

Also make all arguments to getAddress required and document args at call sites.

jamesob commented:
Rebased and addressed @jnewbery's feedback (thanks!):

- Reworked `getAddress` definition and calls to remove use of default params (https://github.com/bitcoin/bitcoin/pull/12830/commits/9db465201d46535a9030d052d858f0e5e2c6e796)
- Moved the `qt/test/util.cpp` makefile inclusion to the right commit (https://github.com/bitcoin/bitcoin/pull/12830/commits/82e8fce0366527fd9f7ab14c32567cf4ef323abd)
- Removed the unused `addressTableModel`

jnewbery commented on src/qt/addresstablemodel.h:
```diff
@@ -12,6 +12,7 @@ enum class OutputType;
 
 class AddressTablePriv;
 class WalletModel;
+class CAddressBookData;
```
forward declaration unused. Please remove

jamesob replied:
Good catch, thanks. Will spend some time figuring out a sensible `clang-tidy` config to catch this sort of stuff.

jnewbery reviewed (commented):
Nearly there!

jnewbery commented:
utACK 41180ac. Good stuff!

jamesob commented on src/qt/editaddressdialog.cpp:
```diff
@@ -129,6 +129,24 @@ void EditAddressDialog::accept()
+            "Address \"%1\" already exists as a receiving address with label "
+            "\"%2\" and so cannot be added as a sending address.";
+    }
+    return tr(fmt_str.c_str()).arg(ui->addressEdit->text()).arg(existing_label);
```
Good catch, fixed.

jamesob commented:
Incorporated @jonasschnelli's good feedback about `tr()` usage. I've also recompiled the EN qt locale with `cd src/ && make translate` and have squashed that in.

jamesob committed:
[qt] Display more helpful message when adding a send address has failed

Addresses #12796.

When we're unable to add a sending address to the address book because it
already exists as a receiving address, display a message indicating as much.
This should help avoid confusion about an address supposedly already in the
book but which isn't currently visible in the interface.

jamesob commented:
Reverted EN locale changes per @jonasschnelli's advice on IRC.

jnewbery commented:
utACK e6e1da33783408182e52fa73a11b6559ee0022f7

jonasschnelli reviewed (approved):
utACK e6e1da33783408182e52fa73a11b6559ee0022f7

jamesob commented:
Any other feedback here?

jimpo commented on src/qt/walletmodel.h:
```diff
@@ -204,6 +204,8 @@ class WalletModel : public QObject
 
     bool isMultiwallet();
+
+    AddressTableModel* getAddressTableModel() const { return addressTableModel; }
```
commit: [tests] [qt] Add tests for address book manipulation via EditAddressDialog

This method is already defined on line 139 but without `const`. I think the non-const variant is actually better because it returns a non-const pointer. (While technically allowed, it feels wrong to me to return a non-const pointer to a member from a const method).

jimpo commented on src/qt/test/addressbooktests.cpp:
```diff
@@ -0,0 +1,143 @@
+    bool firstRun;
+    wallet.LoadWallet(firstRun);
+
+    auto build_address = [&wallet]() {
```
commit: [tests] [qt] Add tests for address book manipulation via EditAddressDialog

nit: Any particular reason for this to be a lambda expression instead of a regular function in the anonymous namespace?

jimpo commented on src/qt/test/addressbooktests.cpp:
```diff
@@ -0,0 +1,143 @@
+        wallet.SetAddressBook(s_key_dest, s_label.toStdString(), "send");
+    }
+
+    auto check_addbook_size = [&wallet](int expected_size) {
```
nit: Also could be a regular function.

jimpo commented on src/qt/test/addressbooktests.cpp:
```diff
@@ -0,0 +1,143 @@
+    check_addbook_size(2);
+
+    // Initialize relevant QT models.
+    std::unique_ptr<const PlatformStyle> platformStyle(PlatformStyle::instantiate("other"));
```
commit: [tests] [qt] Add tests for address book manipulation via EditAddressDialog

nit: camelCase vs snake_case of local variables in this function is inconsistent (`r_key_dest` vs `platformStyle`). I think snake_case is preferred for new code.

jimpo reviewed (commented):
In commit [tests] [qt] Introduce qt/test/util with a generalized ConfirmMessage, new files are missing copyright notice headers.

I tried testing manually. The error message when adding a new address works as expected. However, when editing an existing address book entry and changing the address, the behavior is pretty funky. If the address was previously in the address book, it will close the dialog and do nothing, displaying no error. If it is not in the address book, it will edit the entry but also change the label to the address, despite the label in the form not changing.
